torch/testing/_internal/faulty_rpc_agent/faulty_process_group_agent.cpp

```diff
@@ -0,0 +1,74 @@
+#include <caffe2/torch/testing/_internal/faulty_rpc_agent/faulty_process_group_agent.h>
+
+namespace torch {
+namespace distributed {
+namespace rpc {
+
+std::string fromVec(std::vector<char> vec) {
+  std::string s(vec.begin(), vec.end());
+  return s;
+}
+
+FaultyProcessGroupAgent::FaultyProcessGroupAgent(
+  std::string workerName,
+  std::shared_ptr<c10d::ProcessGroup> pg,
+  int numSendRecvThreads,
+  std::chrono::milliseconds rpcTimeout,
+  std::vector<std::string> messagesToFail,
+  int failNumSends)
+  : ProcessGroupAgent(workerName, pg, numSendRecvThreads, rpcTimeout),
+    failNumSends_(failNumSends) {
+  parseMessagesToFailInput(messagesToFail);
+}
+
+void FaultyProcessGroupAgent::parseMessagesToFailInput(std::vector<std::string> messagesToFail) {
+  // Since we can only pass strings corresponding to the Message Types from the
+  // python tests, we must parse the list of strings and resolve the actual
+  // types. We will then check this list of types in the send function to
+  // determine whether we should fail or not.
+  for (const auto& msgString : messagesToFail) {
+    if (msgString == "RREF_FORK_REQUEST") {
```

Comment:
(don' need to be addressed in this PR)
nit: we can build a const map for this? Not sure if there is a way to do automatic `Enum` <-> `std::string` conversion.

Reply:
Sounds good - I can look into this further in another PR. I tried looking for some automatic `Enum` <> `std::string` conversion (since that would be ideal here), but didn't find clean solutions.